### PR TITLE
Scoping reserved names

### PR DESCRIPTION
**Pull request by kinnrot**
### Summary

Suggestion for resolving #31120 
Don't allow creating scopes named same as ActiveRecord::Relation instance method

### CODE REVIEW

File: activerecord/lib/active_record/scoping/named.rb
Comment (eugeneius) [inline activerecord/lib/active_record/scoping/named.rb]: `Relation.instance_methods` also includes all instance methods defined on `Object`. We shouldn't reject those; declaring a scope named `object_id` works just fine, for example, even if it's an odd thing to do.

I think [`method_defined_within?(name, Relation)`](https://github.com/rails/rails/blob/v5.1.4/activerecord/lib/active_record/attribute_methods.rb#L111) will do the right thing here.

File: activerecord/test/cases/scoping/named_scoping_test.rb
Comment (eugeneius) [inline activerecord/test/cases/scoping/named_scoping_test.rb]: ralation -> relation

File: activerecord/test/cases/scoping/named_scoping_test.rb
Comment (eugeneius) [inline activerecord/test/cases/scoping/named_scoping_test.rb]: If you use `method_defined_within?` as I suggested above, these won't need to change.

**Comment by kinnrot** [comment]
@eugeneius 10x, If you already looked at this code, What do you think about a more actionable message when this issue occurs? How about something that explain's how to cancel auto scope generation from enum or any other places that scopes are auto generated on rail load?

**Comment by eugeneius** [comment]
It looks like `ActiveRecord::Enum` has its own logic to reject dangerous methods: https://github.com/rails/rails/blob/v5.1.4/activerecord/lib/active_record/enum.rb#L219

I think we should extend that to cover instance methods on `Relation` too.

I'm not aware of any other places where Active Record automatically creates scopes for you, so I think adding the enum-specific check should be enough to give users reasonably useful error messages.

**Comment by kinnrot** [comment]
@eugeneius sounds right, I added the check, so what is the way to tell rail to not create a scope for a enum field named 'record' for example ?

**Comment by eugeneius** [comment]
Nice one 😊 could you add a test for the enum validation too?

It's not currently possible to suppress enum scopes; that sounds like a reasonable addition to me, but I think you should wait for a maintainer to review what you have here already and then follow up with a separate pull request afterwards if they're 👍 on the idea.

**Review by rafaelfranca — COMMENTED** [review]
Can you add a CHANGELOG entry and squash your commits after it?

File: activerecord/test/cases/enum_test.rb
Comment (eugeneius) [inline activerecord/test/cases/enum_test.rb]: This should be raising an enum-specific message ("You tried to define an enum named [...]"), which suggests that the new logic there isn't correct.

File: activerecord/lib/active_record/enum.rb
Comment (eugeneius) [inline activerecord/lib/active_record/enum.rb]: When this method is called to detect scope name conflicts, `klass_method` is true:
https://github.com/rails/rails/blob/v5.1.4/activerecord/lib/active_record/enum.rb#L197

A scope is a class method on the model, even if we sometimes access it as an instance method on `Relation`.

Comment (kinnrot) [reply activerecord/lib/active_record/enum.rb]: Ok, So why did the test pass? i

Comment (kinnrot) [reply activerecord/lib/active_record/enum.rb]: Ohh I see, can you show me where is the code that allows accessing the scope via the relation instace?

Comment (eugeneius) [reply activerecord/lib/active_record/enum.rb]: https://github.com/rails/rails/blob/v5.1.4/activerecord/lib/active_record/relation/delegation.rb#L90

File: activerecord/CHANGELOG.md
Comment (eugeneius) [inline activerecord/CHANGELOG.md]: "Fix inconsistency" sounds like we made this work correctly, when really we don't allow it any more. How about:

> Don't allow scopes to be defined which conflict with instance methods on `Relation`.

New changelog entries should be added at the top of the file.

File: activerecord/lib/active_record/enum.rb
Comment (eugeneius) [inline activerecord/lib/active_record/enum.rb]: The error raised here currently says:

> [...] this will generate a instance method "records", which is already defined by ActiveRecord::Relation.

We should pass `type: "class"`, so that it says:

> [...] this will generate a **class** method "records"

Comment (eugeneius) [reply activerecord/lib/active_record/enum.rb]: What do you think of moving this code up by 4 lines, so that both of the `if klass_method` checks are next to each other? It might have been easier to spot the `type: "class"` inconsistency if that was the case.

Comment (eugeneius) [reply activerecord/lib/active_record/enum.rb]: We're referencing `Relation` on the line above, and `ActiveRecord::Relation` here; we can use `Relation` in both places, since this code is inside the `ActiveRecord` module.

File: activerecord/CHANGELOG.md
Comment (kinnrot) [reply activerecord/CHANGELOG.md]: sounds good

File: activerecord/lib/active_record/enum.rb
Comment (kinnrot) [reply activerecord/lib/active_record/enum.rb]: Thanks man!

one question if I may,
When I call a model scope method, Don't I get a relation instance, and when calling a scope method on this relation, it's a class method, where does the delegation happen ?

Comment (eugeneius) [reply activerecord/lib/active_record/enum.rb]: `Relation#method_missing` delegates methods to the model class if it responds to them:
https://github.com/rails/rails/blob/v5.1.4/activerecord/lib/active_record/relation/delegation.rb#L88-L90

File: activerecord/CHANGELOG.md
Comment (eugeneius) [inline activerecord/CHANGELOG.md]: New changelog entries should be added at the top of the file.

File: activerecord/CHANGELOG.md
Comment (eugeneius) [inline activerecord/CHANGELOG.md]: Can you undo the whitespace changes here, and squash your commits again? After that, this looks good to me - thanks for being so responsive to feedback @kinnrot! 🙏

Comment (kinnrot) [reply activerecord/CHANGELOG.md]: is that ok?

Comment (eugeneius) [reply activerecord/CHANGELOG.md]: There's still a diff here, but I'm not sure if it's a blocker for merging - maybe @rafaelfranca can let us know.

**Comment by kinnrot** [comment]
Yeah I tried to take the previous version and not touching the bottom of
the file, but that didn't do the trick

File: activerecord/CHANGELOG.md
Comment (kamipo) [reply activerecord/CHANGELOG.md]: This diff can easily be removed by deleting the four extra whitespaces and adding one blank line.

**Comment by kinnrot** [comment]
The changelog thing is good now.

File: activerecord/CHANGELOG.md
Comment (eugeneius) [inline activerecord/CHANGELOG.md]: So close 😄 but the new changelog entry needs to be _above_ this line, at the very top of the file.

Rails 5.2.0.beta1 has already been released, so it doesn't make sense to include this change under that heading.

**Comment by kinnrot** [comment]
🤗 ohh, I didn't realize that..
well hope to get to beta2

Chen Kinnrot
kinnrot@gmail.com

**Review by eugeneius — APPROVED** [review]

**Comment by hw676018683** [comment]
when I write the following code:

```ruby
  class Attachemnt < ApplicationRecord
    enum extension: {
      unknown: 0,
      image: 1,
      video: 2,
      audio: 3
    }
  end
```

Get the following error:
> You tried to define an enum named "extension" on the model "Attachment", but this will generate a class method "extensions", which is already defined by ActiveRecord::Relation. (ArgumentError)

But I never use something like `Attachment.all.extensions`. So I think it may be more appropriate to only limit scope.
`detect_enum_conflict!(name, name.pluralize, true)`, the class method `name.pluralize` should not be limited.

**Pull request merged**
